src/cxx_support.rs

```diff
@@ -0,0 +1,214 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Support for cxx types.
+
+use std::pin::Pin;
+
+use cxx::{memory::UniquePtrTarget, UniquePtr};
+
+use crate::{Emplace, TryNew};
+
+/// Equivalent of a [`UniquePtr`] which is allocated, but hasn't yet
+/// had its constructor called.
+#[cfg(feature = "cxx")]
+pub struct MaybeUninitUniquePtr<T: UniquePtrTarget>(*mut T);
+
+impl<T: UniquePtrTarget> MaybeUninitUniquePtr<T> {
+  /// Declare that a pointer which we have received from C++
+  /// is, in fact, not initialized but is instead just a blank
+  /// storage area yet to have its constructor called.
+  ///
+  /// Safety: destroying this type results in a memory leak. Callers
+  /// need to take care to either convert this to a normal, fully-
+  /// initialized [`UniquePtr`] by calling [`Self::assume_init`] or
+  /// to manually take care to deallocate the underlying storage.
+  pub unsafe fn from_ptr(ptr: *mut T) -> Self {
+    assert!(!ptr.is_null());
+    Self(ptr)
+  }
+
+  /// Get a pointer to the storage within this uninitialized space.
+  /// This can be used to initialize it.
+  ///
+  /// Safety: callers must call this exactly once per instance, or
+  /// the data may be double-initialized.
+  pub unsafe fn get_uninitialized_space(&mut self) -> Pin<&mut T> {
+    Pin::new_unchecked(self.0.as_mut().unwrap())
+  }
+
+  /// Declare that this [`UniquePtr`] is initialized and has had its
+  /// constructor called.
+  ///
+  /// Safety: callers must ensure that the data was initialized exactly
+  /// once before calling this function.
+  pub unsafe fn assume_init(self) -> UniquePtr<T> {
+    UniquePtr::from_raw(self.0)
+  }
+
+  /// Return the underlying raw pointer such that it can be freed.
+  ///
+  /// Safety: callers must ensure they deallocate the resulting pointer
+  /// to avoid a memory leak. The normal purpose of this method is to
+  /// return a pointer which has been allocated, but not initialized;
+  /// in such a case callers would need to free the memory without calling
+  /// the destructor. However this is not enforced - callers will need
+  /// to free this memory appropriately depending on the lifecycle of this
+  /// object.
+  pub unsafe fn get_for_freeing(self) -> *mut T {
+    self.0
+  }
+}
+
+/// A type which has the ability to create a cxx::UniquePtr which points to
+/// an area of memory with the correct size and alignment, but has not yet had
+/// its constructor called.
+#[cfg(feature = "cxx")]
```

Comment:
If this whole file is dependent on cxx, why do we need to cfg individual items?

(Not to mention that you're otherwise missing cfg attributes on things like impls in this file...)